[GTK][WPE][Skia] Invalid result or crash when painting an accelerated ImageBitmap into WebGL

Scony opened the pull request:
#### b2b8d3a40c38799195f6374ced4e0f4b28dcac17
<pre>
[GTK][WPE][Skia] Invalid result or crash when painting an accelerated ImageBitmap into WebGL
<a href="https://bugs.webkit.org/show_bug.cgi?id=278908">https://bugs.webkit.org/show_bug.cgi?id=278908</a>

Reviewed by Carlos Garcia Campos.

When using skia, transferring ownership of accelerated ImageBitmaps causes GrDirectContext mismatches
that lead to problems with reading pixels, drawing etc. Therefore, we need to let ImageBitmap know
the transfer is happening so that it can act accordingly and e.g. decelerate underlying ImageBuffer.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createImageBuffer):
(WebCore::ImageBitmap::prepareForCrossThreadTransfer):
* Source/WebCore/html/ImageBitmap.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBuffer):
(WebCore::ImageBuffer::sinkIntoImageBufferForCrossThreadTransfer):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::postMessage):
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::postMessage):

Canonical link: <a href="https://commits.webkit.org/283563@main">https://commits.webkit.org/283563@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4aab7fec0414f947038cf756c8c96974f60fc70b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66608 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45983 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19230 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70642 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17741 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68726 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53782 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17503 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53379 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11965 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69675 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42336 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57640 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34045 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39007 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15031 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16095 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60919 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15372 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72344 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14735 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60701 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57711 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61030 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14707 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8692 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2314 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41790 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42867 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44050 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42610 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -175,6 +174,15 @@ std::optional<DetachedImageBitmap> ImageBitmap::detach()
 }
 
+#if USE(SKIA)
+void ImageBitmap::willTransferBetweenThreads()
```
I would use a name to make more explicit that this might change the internal buffer, maybe `prepareForCrossThreadTransfer()` or something like that.

carlosgcampos commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -175,6 +174,15 @@ std::optional<DetachedImageBitmap> ImageBitmap::detach()
+{
+    // FIXME: Try avoiding GPU-to-CPU data transfer by creating accelerated clone that would e.g. leverage underlying GL texture.
+    if (m_bitmap && m_bitmap->renderingMode() == RenderingMode::Accelerated)
+        m_bitmap = m_bitmap->clone(RenderingMode::Unaccelerated);
```
Instead of modifying clone, we could use a new method. There's also `ImageBuffer::sinkIntoBufferForDifferentThread` that only works when the buffer has only one ref for some reason. But we could add our skia specific code there, so here we would just do:

```
m_bitmap = ImageBuffer::sinkIntoBufferForDifferentThread(WTFMove(m_bitmap));
```

and sinkIntoBufferForDifferentThread already takes care of m_bitmap being null or return the same object if we don't need to copy.

carlosgcampos commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -210,15 +210,15 @@ FloatRect ImageBuffer::clampedRect(const FloatRect& rect)
     if (source->resolutionScale() == 1 || preserveResolution == PreserveResolution::Yes) {
         if (source->hasOneRef())
             return source;
     }
```
We should probably check here that renderingMode matches too, at least for skia, because in that case we need the copy even if there's only one ref.

carlosgcampos commented on Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
```diff
@@ -91,7 +95,17 @@ void DedicatedWorkerGlobalScope::prepareForDestruction()
+    // When using skia, transferring ownership of accelerated ImageBitmaps causes GrDirectContext mismatches,
+    // threfore, we need to let ImageBitmap know so that it can act accordingly.
+    for (const auto& transferItem : options.transfer) {
+        if (ImageBitmap* imageBitmap = JSImageBitmap::toWrapped(state.vm(), transferItem.get()))
```
```suggestion
        if (auto* imageBitmap = JSImageBitmap::toWrapped(state.vm(), transferItem.get()))
```

carlosgcampos commented on Source/WebCore/workers/Worker.cpp:
```diff
@@ -141,6 +145,15 @@ Worker::~Worker()
+    // When using skia, transferring ownership of accelerated ImageBitmaps causes GrDirectContext mismatches,
+    // threfore, we need to let ImageBitmap know so that it can act accordingly.
+    for (const auto& transferItem : options.transfer) {
+        if (ImageBitmap* imageBitmap = JSImageBitmap::toWrapped(state.vm(), transferItem.get()))
```
```suggestion
        if (auto* imageBitmap = JSImageBitmap::toWrapped(state.vm(), transferItem.get()))
```

carlosgcampos commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -326,6 +330,16 @@ RefPtr<ImageBuffer> ImageBuffer::sinkIntoBufferForDifferentThread(RefPtr<ImageBu
+    if (!buffer || buffer->renderingMode() != RenderingMode::Accelerated)
+        return buffer;
+    // FIXME: Try avoiding GPU-to-CPU data transfer by creating accelerated clone that would e.g. leverage underlying GL texture.
+    return buffer->clone(RenderingMode::Unaccelerated);
```
Instead of modifying clone, can we just call copyImageBuffer here?

Scony committed:
[GTK][WPE][Skia] Invalid result or crash when painting an accelerated ImageBitmap into WebGL
https://bugs.webkit.org/show_bug.cgi?id=278908

Reviewed by Carlos Garcia Campos.

When using skia, transferring ownership of accelerated ImageBitmaps causes GrDirectContext mismatches
that lead to problems with reading pixels, drawing etc. Therefore, we need to let ImageBitmap know
the transfer is happening so that it can act accordingly and e.g. decelerate underlying ImageBuffer.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/html/ImageBitmap.cpp:
(WebCore::ImageBitmap::createImageBuffer):
(WebCore::ImageBitmap::prepareForCrossThreadTransfer):
* Source/WebCore/html/ImageBitmap.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::copyImageBuffer):
(WebCore::ImageBuffer::sinkIntoImageBufferForCrossThreadTransfer):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/workers/DedicatedWorkerGlobalScope.cpp:
(WebCore::DedicatedWorkerGlobalScope::postMessage):
* Source/WebCore/workers/Worker.cpp:
(WebCore::Worker::postMessage):

Canonical link: https://commits.webkit.org/283563@main
